tsdb/head.go

```diff
@@ -524,23 +531,49 @@ Outer:
 					seriesCreationErr = err
 					break Outer
 				}
-
-				if created {
-					// If this series gets a duplicate record, we don't restore its mmapped chunks,
-					// and instead restore everything from WAL records.
-					series.mmappedChunks = mmappedChunks[series.ref]
+				if !created {
+					// There's already a different ref for this series.
+					multiRef[s.Ref] = series.ref
+				}
+				mmc := mmappedChunks[series.ref]
+				if len(series.mmappedChunks) == 0 && len(mmc) > 0 {
+					series.mmappedChunks = mmc
 
 					h.metrics.chunks.Add(float64(len(series.mmappedChunks)))
 					h.metrics.chunksCreated.Add(float64(len(series.mmappedChunks)))
 
 					if len(series.mmappedChunks) > 0 {
 						h.updateMinMaxTime(series.minTime(), series.maxTime())
 					}
-				} else {
-					// TODO(codesome) Discard old samples and mmapped chunks and use mmap chunks for the new series ID.
+				} else if len(mmc) > 0 && len(series.mmappedChunks) > 0 &&
+					series.mmappedChunks[len(series.mmappedChunks)-1].maxTime > mmc[len(mmc)-1].maxTime {
+					// These new m-mapped chunks in the future for the same series.
+					// Hence replace the old m-mapped chunks with new chunks.
+
+					if overlapsClosedInterval(
+						series.mmappedChunks[0].minTime,
+						series.mmappedChunks[len(series.mmappedChunks)-1].maxTime,
+						mmc[0].minTime,
+						mmc[len(mmc)-1].maxTime,
+					) {
+						// TOOD(codesome) return this instead.
+						panic("overlapping mmapped chunks")
+					}
+
+					h.metrics.chunksCreated.Add(float64(len(mmc)))
+					h.metrics.chunksRemoved.Add(float64(len(series.mmappedChunks)))
+					h.metrics.chunks.Add(float64(len(mmc) - len(series.mmappedChunks)))
+
+					series.mmappedChunks = mmc
+
+					if series.headChunk != nil && mmc[len(mmc)-1].maxTime >= series.headChunk.minTime {
+						// The head chunk was completed and was m-mapped after taking the snapshot.
```

Comment:
again, why this would happen? Why something could be mapped after taking (chunk) snapshot? This can be done only on shutdown

Reply:
Currently, the code considers snapshot at intervals, hence this was possible. Once we can confirm there is no race during shutdown where it can happen, we can remove this check. (Race of a sample coming in after snapshot which causes the chunk to m-map).

Reply:
:+1:  Let's commen that

Reply:
Another reason why this could happen: Prometheus shutdown ungracefully and restart led it to pick up the old snapshot, hence we have a old chunk which needs to be cleaned up here. Will add that in the comments.

Reply:
It will go in https://github.com/prometheus/prometheus/pull/7438